[feat] added split dataset feature

ytsheng opened the pull request:
* allow user to specify x% of dataset to split from train, example yaml shown below:
```
  textvqa:
    zoo_requirements:
    - textvqa.defaults
    - textvqa.ocr_en
    split_train:
      val: 0.2
      test: 0.1
      seed: 42
    features:
      train:
      - textvqa/defaults/features/open_images/detectron.lmdb,textvqa/ocr_en/features/ocr_en_frcn_features.lmdb
   processors:
```
* To test: `pytests test`
* See logs, 24221 is 0.7 of 34602 which is the total size of the train after calculation, etc.
![Screen Shot 2020-08-10 at 11 47 12 PM](https://user-images.githubusercontent.com/1545487/89866133-e052f600-db63-11ea-80e0-4f2c6b1a7acc.png)

ytsheng commented on mmf/datasets/subset_dataset.py:
```diff
@@ -0,0 +1,15 @@
+from torch.utils.data.dataset import Subset
```
oh I need to add a comment on copyright... Will do.

ytsheng commented:
hold on the review for a moment, running into some git push issues.

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -119,6 +123,49 @@ def _download_requirement(
         self.config = config
+
+        new_config = config
+        split_dataset_from_train = self.config.split_train
```
This will break existing datasets, use 
`self.config.get("split_train", False)`

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -119,6 +123,49 @@ def _download_requirement(
+            if data_type == "seed":
+                continue
+            if config.use_images:
+                modified_config.images[data_type] = config.images.train
```
Use deepcopy to assign as well.

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -119,6 +123,49 @@ def _download_requirement(
+        self.dataset = dataset
+        return self.dataset
+
+    def _modify_dataset_config(self, config):
```
Make this function's name more specific. `_modify_dataset_config_for_split`

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -133,17 +180,25 @@ def load(self, config, dataset_type, *args, **kwargs):
+        if start >= 1.0:
+            raise ValueError(
+                "Ratios of val and test should not exceed 100%."
+                + "Need to leave some percentage for training."
```
```suggestion
                + " Need to leave some percentage for training."
```

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -133,17 +180,25 @@ def load(self, config, dataset_type, *args, **kwargs):
+            start += self.config.split_train[data_type]
+
+        if start >= 1.0:
+            raise ValueError(
```
This probably should be a warning in case of == 1.0 as I can launch this on val inference alone.

ytsheng replied:
added on the bottom

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -119,6 +123,49 @@ def _download_requirement(
+        split_dataset_from_train = self.config.split_train
+        if split_dataset_from_train:
+            split_range = self._calculate_split_for_dataset_type(dataset_type)
+            new_config = self._modify_dataset_config(config)
```
Why not directly assign to config?

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -119,6 +123,49 @@ def _download_requirement(
+        if annotations is None:
+            return None
+
+        datasets = []
+        for imdb_idx in range(len(annotations)):
+            dataset_class = self.dataset_class
+            dataset = dataset_class(new_config, dataset_type, imdb_idx)
+            datasets.append(dataset)
+
+        dataset = MMFConcatDataset(datasets)
+        if split_dataset_from_train:
+            start, end = split_range
+            dataset_length = len(dataset)
+            start, end = round(start * dataset_length), round(end * dataset_length)
+            seed = new_config.split_train.seed
+            indices = np.random.RandomState(seed).permutation(dataset_length)[start:end]
+            dataset = MMFSubset(dataset, indices)
```
We can probably move this to a separate function at this point

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -119,6 +123,49 @@ def _download_requirement(
+            dataset_length = len(dataset)
+            start, end = round(start * dataset_length), round(end * dataset_length)
+            seed = new_config.split_train.seed
+            indices = np.random.RandomState(seed).permutation(dataset_length)[start:end]
```
Use torch to generate indices rather than np?

ytsheng replied:
https://github.com/pytorch/pytorch/issues/17079 
It seems like the numpy interface is much cleaner compared to torch's.
Is there a reason that we prefer to use torch than numpy?

apsdehal replied:
Yeah, one extra thing for user to seed. We try to rely on torch as much as possible instead of numpy.

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -119,6 +123,49 @@ def _download_requirement(
+        if split_dataset_from_train:
+            start, end = split_range
+            dataset_length = len(dataset)
+            start, end = round(start * dataset_length), round(end * dataset_length)
```
Make sure there is no overlap between two sets here.

ytsheng replied:
This only calculates the start, end index for one dataset_type (ie., val, test, train).

apsdehal replied:
Yes, i understand that, but they can overlap given the way they are calculated.

apsdehal commented:
Formatting is failing on CircleCI, please fix that as well.

ytsheng commented:
Thanks for the review. I addressed your comments. I had to make some changes in the multi_dataset_loader due to the way empty dataset is handled (ie, if the val takes up 100% of the train, then the train will be of length 0).

apsdehal commented on mmf/datasets/multi_dataset_loader.py:
```diff
@@ -132,17 +134,28 @@ def build_datasets(self, config):
 
     def build_dataloaders(self):
-        assert len(self._datasets) > 0, "Call build_datasets first"
+        if not self._did_build_datasets:
```
Why do we need this extra check? Maybe I am missing something here?

apsdehal commented on mmf/datasets/multi_dataset_loader.py:
```diff
@@ -176,9 +189,14 @@ def _infer_dataset_probabilities(self):
+            dataset_lengths = self._per_dataset_lengths[:]
+            dataset_lengths = [
+                # Empty datasets need to have some probability to prevent infinite loop
+                max(1, dataset_len)
```
If they are already popped how would they reach this stage?

apsdehal commented on mmf/datasets/multi_dataset_loader.py:
```diff
@@ -132,17 +134,28 @@ def build_datasets(self, config):
+        if not self._did_build_datasets:
+            assert len(self._datasets) > 0, "Call build_datasets first"
+
+        if len(self.datasets) == 0:
```
I think this logic is getting unnecessarily complicated for a single use case. We should fail fast when there is no datasets, or use try except python style.

apsdehal reviewed (changes requested):
Code readability has been impacted with the recent changes, it is hard to tell what is happening. See my comment on how to improve. Since, the feature for 0 size dataset is getting too complex for a simple feature, we should fail fast and prevent this. We shouldn't  avoid making these many changes in MultiDatasetLoader as it is most sensitive part of MMF.

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -119,6 +123,61 @@ def _download_requirement(
+                    f"Train split ratio for {dataset_type} must be positive."
+                )
+
+            generator = torch.Generator()
+            generator.manual_seed(self.config.get("split_train.seed", 123456))
+            indices = torch.randperm(dataset_length, generator=generator)[start:end]
```
NIT: shift this to a separate function

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -119,6 +123,61 @@ def _download_requirement(
+            if data_type == "seed":
+                continue
+            if config.use_images:
+                with open_dict(config.images):
```
NIT: You can open dict once on top.

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -119,6 +123,61 @@ def _download_requirement(
+                continue
+            if config.use_images:
+                with open_dict(config.images):
+                    config.images[data_type] = copy.deepcopy(config.images.train)
```
NIT: Do `from copy import deepcopy`

apsdehal commented on mmf/datasets/multi_dataset_loader.py:
```diff
@@ -132,17 +134,28 @@ def build_datasets(self, config):
+                    self._total_length += dataset_instance_length
+                else:
+                    # Filter out datasets with no items
+                    self.datasets.pop()
```
assert and fail here fast.

ytsheng commented:
Hi @apsdehal, thanks for the review! I addressed your comments and ran the tests. Here is the documentation: -

We explicitly assert when the dataset length is of size 0, this is demonstrated by the following dataset_config:
```
  textvqa:
    zoo_requirements:
    - textvqa.defaults
    - textvqa.ocr_en
    split_train:
      val: 0.9
      test: 0.1
    features:
      train:
      - textvqa/defaults/features/open_images/detectron.lmdb,textvqa/ocr_en/features/ocr_en_frcn_features.lmdb
   processors:
```
Here, we assert due to val + test ratio add up to 1, leaving 0 for training. However if you wish to convert most of train data to validation, you can do the following:
```
  textvqa:
    zoo_requirements:
    - textvqa.defaults
    - textvqa.ocr_en
    split_train:
      val: 0.99
      test: 0.001
    features:
      train:
      - textvqa/defaults/features/open_images/detectron.lmdb,textvqa/ocr_en/features/ocr_en_frcn_features.lmdb
      test:
      - test_path
   processors:
```
Here if you specified test's train split percentage, then we will use the specified percentage from train for test. However, if you don't specify test's train split percentage, we will attempt to use the data from the specified test_path. If there is no specified test_path and you don't specify the train_split percentage, the code will raise a ValueError.


I will add this new API change into the docs.

apsdehal commented on mmf/datasets/multi_dataset_loader.py:
```diff
@@ -132,8 +132,10 @@ def build_datasets(self, config):
-                self._per_dataset_lengths.append(len(dataset_instance))
-                self._total_length += len(dataset_instance)
+                dataset_instance_length = len(dataset_instance)
+                assert dataset_instance_length
```
be more specific on why this assert failed with a message.

apsdehal commented on mmf/datasets/mmf_dataset_builder.py:
```diff
@@ -119,6 +122,64 @@ def _download_requirement(
+        return dataset
+
+    def _generate_permuted_indexes(self, dataset_length):
+        import torch
```
Move to top.

apsdehal reviewed (approved):
This looks good to me to land. Please address minor comments before landing and update the API docs.

apsdehal commented:
Also, make sure to reimport the diff to phabrictor before shipping it off.
